test/metabase/lib/datascript_test.cljc

```diff
@@ -0,0 +1,1100 @@
+                                                         :metadata.table/name "ORDERS"}}}
+              query))
+      (testing "has only one stage"
+        (is (= (:mbql.query/stage0 query)
+               (:mbql.query/stage  query))))))
+
+  (testing "filters"
+    (testing "one filter"
+      (let [query (-> (ld/query meta/metadata-provider (meta/table-metadata :orders))
+                      (ld/filter -1 (ld/expr :< [:metadata.field/id (meta/id :orders :subtotal)] 100)))
+            exp   {:mbql.clause/operator :<
+                   :mbql.clause/argument
+                   [{:mbql/ref    {:metadata.field/id (meta/id :orders :subtotal)}
+                     :mbql/series 0}
+                    {:mbql/lit    100
+                     :mbql/series 1}]
+                   :mbql/series 0}]
+        (is (=? {:mbql.query/stage {:mbql.stage/filter [exp]}}
+                query))
+        (is (=? [exp]
+                (ld/filters query -1))))
+      (testing "nested expressions"
+        (let [query (-> (ld/query meta/metadata-provider (meta/table-metadata :orders))
+                        (ld/filter -1 (ld/expr :>
+                                               (ld/expr (keyword "/")
+                                                        [:metadata.field/id (meta/id :orders :discount)]
+                                                        [:metadata.field/id (meta/id :orders :subtotal)])
+                                               0.1)))
+              exp   {:mbql.clause/operator :>
+                     :mbql.clause/argument
+                     [{:mbql/sub    {:mbql.clause/operator (keyword "/")
+                                     :mbql.clause/argument
+                                     ;; Showing that the order here doesn't matter.
+                                     [{:mbql/ref {:metadata.field/id (meta/id :orders :subtotal)}
+                                       :mbql/series 1}
+                                      {:mbql/ref {:metadata.field/id (meta/id :orders :discount)}
+                                       :mbql/series 0}]}
+                       :mbql/series 0}
+                      {:mbql/lit    0.1
+                       :mbql/series 1}]
+                     :mbql/series 0}]
+          (is (=? {:mbql.query/stage {:mbql.stage/filter [exp]}}
+                  query))
+          (is (=? [exp]
+                  (ld/filters query -1))))))
+    (testing "multiple filters"
+      (let [query            (-> (ld/query meta/metadata-provider (meta/table-metadata :orders))
+                                 (ld/filter -1 (ld/expr :> [:metadata.field/id (meta/id :orders :subtotal)] 100))
+                                 ;; START HERE: I need to think about implicit joins a bit more carefully.
+                                 ;; Implicit joins are really at the column level and the source level.
+                                 ;; They need some indirection.
```

Comment:
Whoops, there's a note to myself that survived to the PR stage!

I have thought more about implicit joins since! They're used below in some tests.